internal/controllers/machineset/machineset_controller.go

```diff
@@ -635,8 +637,19 @@ func (r *Reconciler) getMachineSetsForMachine(ctx context.Context, m *clusterv1.
 	return mss, nil
 }
 
+// shouldAdopt returns true if the MachineSet should be adopted as a stand-alone MachinesSet directly owned by the Cluster.
 func (r *Reconciler) shouldAdopt(ms *clusterv1.MachineSet) bool {
-	return !util.HasOwner(ms.OwnerReferences, clusterv1.GroupVersion.String(), []string{"MachineDeployment", "Cluster"})
+	// if the machine set is not yet controller by a machine deployment, or if it is not already been adopted as a stand-alone MachinesSet directly owned by the Cluster.
+	if !util.HasOwner(ms.OwnerReferences, clusterv1.GroupVersion.String(), []string{"MachineDeployment", "Cluster"}) {
+		// If the MachineSet is originated by a MachineDeployment, prevents it to be adopted as a stand alone MachineSet.
+		// Note: this is required because after restore from a backup both the MachineSet controller and the
+		// MachineDeployment controller are racing to adopt MachineSets, see https://github.com/kubernetes-sigs/cluster-api/issues/7529
```

Comment:
```suggestion
		// If the MachineSet is originated by a MachineDeployment, this prevents from being adopted as a stand alone MachineSet.
		// Note: this is required because after restore from a backup both the MachineSet controller and the
		// MachineDeployment controller are racing to adopt MachineSets, see https://github.com/kubernetes-sigs/cluster-api/issues/7529
```